[DOC-107] add a page for Data types in Explore YCQL features (#18942)
yugabyte/yugabyte-db: docs/content/preview/explore/ycql-language/data-types.md

---
title: Data types
linkTitle: Data types
description: Data types in YCQL
image: /images/section_icons/secure/create-roles.png
menu:
  preview:
    identifier: explore-ycql-language-features-data-types
    parent: explore-ycql-language
    weight: 150
type: docs
---

This page describes the data types supported in YCQL, from the basic data types to collections, and user defined types.

The [JSONB document data type](../../json-support/jsonb-ycql/) is described in a separate section.

{{% explore-setup-single %}}

## Strings

The following character types are supported:

* `varchar`: String of unicode characters of unlimited length
* `text`: String of unicode characters of unlimited length

`varchar` and `text` are aliases.

The following Apache Cassandra character types are not supported:

* `ascii`: Use `text` or `varchar`

To test YugabyteDB support for character types, create a table that has columns with the following types specified:

> ```sql

Author: Good catch. My bad. I was use an existing keyspace.

```sql
CREATE TABLE albums (
  album_id BIGINT PRIMARY KEY,
  title VARCHAR,
  play_time SMALLINT,
  library_record INT
);

INSERT INTO albums (album_id, title, play_time, library_record)
values (3223372036854775808,'Funhouse', 3600,2146483645 );
```

Similarly, the following example shows how to create a table with floating-point typed columns and how to insert a row into that table:

```sql
CREATE TABLE floating_point_test (
  float_test FLOAT PRIMARY KEY,
  decimal_test DECIMAL
);

INSERT INTO floating_point_test (float_test, decimal_test)
VALUES (92233720368547.75807, 5.36152342);
```

## Date and time

Temporal data types allow us to store date and time data. The following date and time types are supported in PostgreSQL and YugabyteDB:

* `DATE`: stores the dates only
* `TIME`: stores the time of day values with nanosecond precision
* `TIMESTAMP`: stores both date and time values with milliseconds precision

The following example creates a table with the temporal types:

```sql
CREATE TABLE temporal_types (
  date_type DATE PRIMARY KEY,
  time_type TIME,
  timestamp_type TIMESTAMP,
);
```

The following example inserts a row into the table:

> ```sql

Author: Looks like i didn't test these very well. Fixed now.

 28df63b7-cc57-43cb-9752-fae69d1653da |  John Doe | ['2023-02-03 04:05:00.000000+0000'] | {'home': '669-555-1212', 'work': '408-555-2121'} |  {'sa-1011212'}

(1 rows)
```

When the user logs in again, the 'recent_logins' LIST column can be updated as shown below:

```sql
UPDATE user_profile
SET recent_logins = recent_logins + ['2023-04-05 09:15:08.000000+0000']
WHERE user_id = 28df63b7-cc57-43cb-9752-fae69d1653da;
```

```sql
ycqlsh> select * from user_profile;
```

```output
 user_id                              | user_name | recent_logins                                                          | phone_numbers                                    | account_numbers
--------------------------------------+-----------+------------------------------------------------------------------------+--------------------------------------------------+-----------------
 28df63b7-cc57-43cb-9752-fae69d1653da |  John Doe | ['2011-02-03 04:05:00.000000+0000', '2023-04-05 09:15:08.000000+0000'] | {'home': '669-555-1212', 'work': '408-555-2121'} |  {'sa-1011212'}

(1 rows)
```

> The above example appends the new element to an existing list. Prepending is also possible as shown

Reviewer: ```suggestion
The preceding example appends the new element to an existing list. Prepending is also possible, as follows:
```

> below:

Reviewer: ```suggestion
```

```sql
UPDATE user_profile
SET recent_logins = ['2023-04-05 09:15:08.000000+0000'] + recent_logins
WHERE user_id = 28df63b7-cc57-43cb-9752-fae69d1653da;
```

```sql
ycqlsh> select * from user_profile;
```

```output
 user_id                              | user_name | recent_logins                                                          | phone_numbers                                    | account_numbers
--------------------------------------+-----------+------------------------------------------------------------------------+--------------------------------------------------+-----------------
 28df63b7-cc57-43cb-9752-fae69d1653da |  John Doe | ['2023-04-05 09:15:08.000000+0000', '2011-02-03 04:05:00.000000+0000'] | {'home': '669-555-1212', 'work': '408-555-2121'} |  {'sa-1011212'}

(1 rows)
```

> Working with `SET` and `MAP` is similar to the above example except the these types do not have a

Reviewer: ```suggestion
`SET` and `MAP` work similarly, except the these types do not have a
```

notion of prepending and the syntax for literals is slightly different. See [YCQL
Collections](../../../api/ycql/type_collection) for more details.

## User defined types

A user defined type is a collection of data types similar to a `struct` in a programming language.

> ### 1. Create a composite type

Reviewer: Steps shouldn't be separate headings.

> ### 1. Create a composite type

Reviewer: ```suggestion
1. Create a composite type as follows:
```

```sql
CREATE TYPE inventory_item (
   name text,
   supplier_id integer,
   price float
);
```

> ### 2. Create table with a composite type

Reviewer: ```suggestion
1. Create a table with a composite type as follows:
```

```sql
CREATE TABLE on_hand (
   item_id UUID PRIMARY KEY,
   item inventory_item,
   count integer
);
```

> ### 3. Insert a row

Reviewer: ```suggestion
```

> To insert a row as follows:

Reviewer: ```suggestion
1. Insert a row as follows:
```

```sql
INSERT INTO on_hand (item_id, item, count) VALUES (28df63b7-cc57-43cb-9752-fae69d1653da, {name: 'fuzzy dice', supplier_id: 42, price: 1.99}, 1000);
```

> ### 4. Select data

Reviewer: ```suggestion
```

> To select data from the `on_hand` example table, execute the following:

Reviewer: ```suggestion
1. To select data from the `on_hand` example table, execute the following:
```

follows:

```sql
UPDATE user_profile
SET recent_logins = ['2023-04-05 09:15:08.000000+0000'] + recent_logins
WHERE user_id = 28df63b7-cc57-43cb-9752-fae69d1653da;
```

```sql
ycqlsh> select * from user_profile;
```

```output
 user_id                              | user_name | recent_logins                                                          | phone_numbers                                    | account_numbers
--------------------------------------+-----------+------------------------------------------------------------------------+--------------------------------------------------+-----------------
 28df63b7-cc57-43cb-9752-fae69d1653da |  John Doe | ['2023-04-05 09:15:08.000000+0000', '2011-02-03 04:05:00.000000+0000'] | {'home': '669-555-1212', 'work': '408-555-2121'} |  {'sa-1011212'}

(1 rows)
```

`SET` and `MAP` work similarly, except that these types do not have a
notion of prepending and the syntax for literals is slightly different. See [YCQL
Collections](../../../api/ycql/type_collection) for more details.

## User defined types

A user defined type is a collection of data types similar to a `struct` in a programming language.

1. Create a user defined type

> ```sql

Author: Ah! i was wondering why you had all 1s in ur suggested changes earlier.

---
title: Data types
linkTitle: Data types

> description: Data types in YSQL

Reviewer: Looks like the YSQL page, not YCQL?

> description: Data types in YSQL

Reviewer: ```suggestion
description: Data types in YCQL
```

image: /images/section_icons/secure/create-roles.png
menu:
  preview:

>     identifier: explore-ysql-language-features-data-types

Reviewer: ```suggestion
    identifier: explore-ycql-language-features-data-types
```

>     parent: explore-ysql-language-features

Reviewer: ```suggestion
    parent: explore-ycql-language
```

    weight: 150
type: docs
---

> This document describes the data types supported in YSQL, from the basic data types to the `SERIAL` pseudo-type (for implementing an auto-incrementing primary key column in a table), arrays, composite types, and range types.

Reviewer: wrong content?
